src/java/org/apache/cassandra/index/sai/disk/v1/InvertedIndexSearcher.java

```diff
@@ -129,11 +152,105 @@ else if (exp.getOp() == Expression.Op.RANGE)
         throw new IllegalArgumentException(indexContext.logMessage("Unsupported expression: " + exp));
     }
 
+    private Cell<?> readColumn(SSTableReader sstable, PrimaryKey primaryKey)
+    {
+        var dk = primaryKey.partitionKey();
+        var slices = Slices.with(indexContext.comparator(), Slice.make(primaryKey.clustering()));
+        try (var rowIterator = sstable.iterator(dk, slices, columnFilter, false, SSTableReadsListener.NOOP_LISTENER))
+        {
+            var unfiltered = rowIterator.next();
+            assert unfiltered.isRow() : unfiltered;
+            Row row = (Row) unfiltered;
+            return row.getCell(indexContext.getDefinition());
+        }
+    }
+
     @Override
     public CloseableIterator<PrimaryKeyWithSortKey> orderBy(Orderer orderer, Expression slice, AbstractBounds<PartitionPosition> keyRange, QueryContext queryContext, int limit) throws IOException
     {
-        var iter = new RowIdWithTermsIterator(reader.allTerms(orderer.isAscending()));
-        return toMetaSortedIterator(iter, queryContext);
+        if (!orderer.isBM25())
+        {
+            var iter = new RowIdWithTermsIterator(reader.allTerms(orderer.isAscending()));
+            return toMetaSortedIterator(iter, queryContext);
+        }
+
+        // find documents that match each term
+        var queryTerms = orderer.extractQueryTerms();
+        var postingLists = queryTerms.stream()
+                                     .collect(Collectors.toMap(Function.identity(),  term ->
+                                     {
+                                         var encodedTerm = version.onDiskFormat().encodeForTrie(term, indexContext.getValidator());
+                                         var listener = MulticastQueryEventListeners.of(queryContext, perColumnEventListener);
+                                         return reader.exactMatch(encodedTerm, listener, queryContext);
+                                     }));
+        // extract the match count for each
+        var documentFrequencies = postingLists.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> (long) e.getValue().size()));
+
+        try (var pkm = primaryKeyMapFactory.newPerSSTablePrimaryKeyMap();
+             var merged = IntersectingPostingList.intersect(List.copyOf(postingLists.values())))
```

Comment:
What is the motivation for using intersection here as opposed to union? I know that our `:` operator does intersection, but I think that is the opposite of the typical expectation for analyzed text search. Since BM25 is its own thing, I wonder if we can break with our previous design decision and union these results. I am assuming that docs with more terms will naturally be higher in the list, anyway.

Reply:
(1) we're materializing the results into memory and there's no real way around that so union increases the risk that something explodes

(2) I'm relying on "Our experiments focused on conjunctive query evaluation, where the document must contain all query terms; previous work [1] has shown that this approach yields comparable end-to-end effectiveness to disjunctive query evaluation, but is faster." in https://dl.acm.org/doi/abs/10.1145/2600428.2609460 [the work cited is by one of the authors so I assume it's accurate]

Reply:
> we're materializing the results into memory and there's no real way around that

actually we can avoid this (while still doing a single pass) if we materialize average document length in the statistics